src/Frame.cc

```diff
@@ -402,18 +403,55 @@ bool Frame::FillRasterTileData(CARTA::RasterTileData& raster_tile_data, const Ti
             return true;
         } else if (compression_type == CARTA::CompressionType::ZFP) {
             auto nan_encodings = GetNanEncodingsBlock(tile_image_data, 0, tile_width, tile_height);
-            tile_ptr->set_nan_encodings(nan_encodings.data(), sizeof(int32_t) * nan_encodings.size());
 
             if (ZStokesChanged(z, stokes)) {
                 return false;
             }
 
             auto t_start_compress_tile_data = std::chrono::high_resolution_clock::now();
-            std::vector<char> compression_buffer;
-            size_t compressed_size;
-            int precision = lround(compression_quality);
-            Compress(tile_image_data, 0, compression_buffer, compressed_size, tile_width, tile_height, precision);
-            tile_ptr->set_image_data(compression_buffer.data(), compressed_size);
+            {
+                // compress the data with a default precision
+                std::vector<char> compression_buffer;
+                size_t compressed_size;
+                int precision = lround(compression_quality);
+                Compress(tile_image_data, 0, compression_buffer, compressed_size, tile_width, tile_height, precision);
+                float compression_ratio = (float)compressed_size / (float)tile_image_data_size;
+
+                if (precision < HIGH_COMPRESSION_QUALITY && compression_ratio < 0.05) {
+                    // re-compress the data with a higher precision
+                    std::vector<char> compression_buffer_2;
+                    size_t compressed_size_2;
+                    Compress(
+                        tile_image_data, 0, compression_buffer_2, compressed_size_2, tile_width, tile_height, HIGH_COMPRESSION_QUALITY);
+                    float compression_ratio_2 = (float)compressed_size_2 / (float)tile_image_data_size;
+
+                    if (compression_ratio_2 < 0.1) {
+                        // set compression data with high precision
+                        raster_tile_data.set_compression_quality(HIGH_COMPRESSION_QUALITY);
+                        tile_ptr->set_nan_encodings(nan_encodings.data(), sizeof(int32_t) * nan_encodings.size());
+                        tile_ptr->set_image_data(compression_buffer_2.data(), compressed_size_2);
+                        spdlog::debug(
+                            "Change to high compression quality: {}->{}. The change of compression ratio for tile (layer:{}, x:{}, y:{}) "
```

Comment:
I suggest splitting this into the normal message about the ratio and an additional message which says something like:

"Using high compression quality. Previous compression ratio: {:.3f}"

I don't think it's necessary to mention the low precision because we don't normally, or the high precision because it's a constant.

Reply:
The zfp precision here can be even greater than 32. 32 here is our choice of the upper limit we provide to users. It would be still good to mention we change from which lower precision to what we selected as high precision in the debug message just in case in the future we might need it for debugging purpose.

Reply:
Fair enough; I'll edit my comment.